File: examples/core/representative_selection_space.cpp
#include <metric/metric/catalog.hpp>
#include <metric/space/index/operators.hpp>
#include <metric/space.hpp>

#include <cassert>
#include <iostream>
#include <string>
#include <vector>

int main()
{
	std::vector<std::string> records = {"cat", "cot", "coat", "dog"};
	const auto space = mtrc::Space::from_records(records, mtrc::Edit<std::string>{});
	const auto selected = mtrc::stats::structural_analysis::representative_indices(records, mtrc::Edit<std::string>{}, 3);
	const auto selected_records = mtrc::stats::structural_analysis::representatives(records, mtrc::Edit<std::string>{}, 3);
	const auto center = mtrc::stats::structural_analysis::medoid_index(records, mtrc::Edit<std::string>{});
	const auto center_record = mtrc::stats::structural_analysis::medoid(records, mtrc::Edit<std::string>{});
	const auto separated = mtrc::stats::structural_analysis::separated_representative_indices(records, mtrc::Edit<std::string>{}, 2);
	const auto separated_records =
		mtrc::stats::structural_analysis::separated_representatives(records, mtrc::Edit<std::string>{}, 2);
	const auto covered = mtrc::stats::structural_analysis::coverage_representative_indices(records, mtrc::Edit<std::string>{}, 1);
	const auto covered_records = mtrc::stats::structural_analysis::coverage_representatives(records, mtrc::Edit<std::string>{}, 1);

	assert((selected == std::vector<std::size_t>{0, 3, 1}));
	assert((selected_records == std::vector<std::string>{"cat", "dog", "cot"}));
	assert(center == 1);
	assert(center_record == "cot");
	assert((separated == std::vector<std::size_t>{0, 3}));
	assert((separated_records == std::vector<std::string>{"cat", "dog"}));
	assert((covered == std::vector<std::size_t>{0, 3}));
	assert((covered_records == std::vector<std::string>{"cat", "dog"}));
	assert(space.distance(selected[0], selected[1]) == 3);

	std::cout << "representatives:";
	for (const auto index : selected) {
		std::cout << " " << records[index];
	}
	std::cout << "\n";
	std::cout << "medoid representative: " << records[center] << "\n";
	std::cout << "separated representatives:";
	for (const auto index : separated) {
		std::cout << " " << records[index];
	}
	std::cout << "\n";
	std::cout << "radius-cover representatives:";
	for (const auto index : covered) {
		std::cout << " " << records[index];
	}
	std::cout << "\n";
	std::cout << "distance(cat, dog) = " << space.distance(selected[0], selected[1]) << "\n";

	return 0;
}

File: python/examples/metric_space/representative_selection_space.py
"""Representative selection over a transport space.

Representative selection (farthest-first, medoid, separated, coverage) is a
METRIC algorithm. The Python facade delegates these selectors to the native C++
binding and returns deterministic representative IDs / records while keeping the
records in their original histogram form.
"""

from metric import (
    Space,
    coverage_representative_indices,
    coverage_representatives,
    medoid,
    medoid_index,
    representative_indices,
    representatives,
    separated_representative_indices,
    separated_representatives,
)


def cumulative_transport_distance(lhs, rhs):
    """One-dimensional earth mover distance for equal-mass histograms."""
    if len(lhs) != len(rhs):
        raise ValueError("histograms must have the same number of bins")

    cumulative_delta = 0.0
    distance = 0.0
    for lhs_mass, rhs_mass in zip(lhs, rhs):
        cumulative_delta += lhs_mass - rhs_mass
        distance += abs(cumulative_delta)

    return distance


def main():
    names = ["left-edge", "one-step", "right-edge", "split-left", "center"]
    records = [
        (1.0, 0.0, 0.0, 0.0),
        (0.0, 1.0, 0.0, 0.0),
        (0.0, 0.0, 0.0, 1.0),
        (0.5, 0.5, 0.0, 0.0),
        (0.0, 0.5, 0.5, 0.0),
    ]

    space = Space(records, cumulative_transport_distance)

    # Adapter surface: explicit distances over a caller-provided metric.
    assert space.distance(0, 2) == 3.0
    print("records =", ", ".join(names))
    print("distance(left-edge, right-edge) =", space.distance(0, 2))

    metric = cumulative_transport_distance
    selected = representative_indices(records, metric, k=3)
    assert selected == (0, 2, 4)
    assert representatives(records, metric, k=3) == tuple(records[index] for index in selected)
    assert medoid_index(records, metric) == 1
    assert medoid(records, metric) == records[1]
    assert separated_representative_indices(records, metric, 1.5) == (0, 2, 4)
    assert separated_representatives(records, metric, 1.5) == (records[0], records[2], records[4])
    assert coverage_representative_indices(records, metric, 1.5) == (0, 2)
    assert coverage_representatives(records, metric, 1.5) == (records[0], records[2])

    print("farthest-first representatives =", [names[index] for index in selected])
    print("medoid =", names[medoid_index(records, metric)])


if __name__ == "__main__":
    main()
